### PR TITLE
Audio: Convert Google RTC audio processing to module API

### PR DESCRIPTION
**Pull request by singalsu**
Fist commit is module adapter conversion

Second commit is improvement to mockup to mix and saturate all available matching channels.

Third commit adds build and load to testbench.

Related to, no more dependent of:

- [ ] https://github.com/thesofproject/sof/pull/8237

### CODE REVIEW

**Review by lgirdwood — COMMENTED** [review]
@singalsu good stuff !
@johnylin76 @cujomalainey fyi - pls review.

**Comment by singalsu** [comment]
@cujomalainey This is the first version that works with created test topology in sof-nocodec style, sharing it soon as well. I think it's best to get #8237 merged first, then I'll rebase this.

**Comment by singalsu** [comment]
Fixed conflict with 7112246d4fef43e3a5caffbbd0c1d58f14c02303 ("google_rtc_audio_processing: remove buffer_acquire")

File: src/audio/google/google_rtc_audio_processing.c
Comment (lyakh) [inline src/audio/google/google_rtc_audio_processing.c:337]: in principle `.copy()` can run concurrently with IPCs, in which case the timer interrupt can interrupt IPC processing, then the scheduler will choose between the IPC thread and the LL scheduler thread and will pick up the latter, since it has a higher priority. But if the component has only one sink and one source, mostly it's only the "stop" IPC that can arrive while the stream is already running, or can some configuration be sent to a running stream too?

**Comment by RanderWang** [comment]
@singalsu  How do you test it ? it is used by chrome so it is better to sync with chrome team.

**Comment by singalsu** [comment]
> @singalsu How do you test it ? it is used by chrome so it is better to sync with chrome team.

I tested this in UPX nocodec mode with new very topology in PR #8266. Testing with Chrome sounds like a good idea but I can't do it myself. I'll see if I can get help from them to test this.

**Comment by yongzhi1** [comment]
Sanity checked the patch along with https://github.com/thesofproject/linux/pull/4550 and [#8101](https://github.com/thesofproject/sof/pull/8101), [#8246](https://github.com/thesofproject/sof/pull/8246), [#8204](https://github.com/thesofproject/sof/pull/8204), no regression found with AEC mock module.

File: src/audio/google/google_rtc_audio_processing.c
Comment (singalsu) [reply src/audio/google/google_rtc_audio_processing.c:337]: I preserved this part as close to the original so this should not change the operation. @lkoenig What do you think of this?

Comment (lkoenig) [reply src/audio/google/google_rtc_audio_processing.c:337]: Configurations can be sent when a stream is running.

Comment (singalsu) [reply src/audio/google/google_rtc_audio_processing.c:337]: @lyakh @lkoenig I think it depends on the (unknown) internals of the processing whether new config data blob during copy() is a hazard  or not. The open-source mock-up version ignores the blob content totally.

The reconfigure set here should ensure that the next copy() correctly re-initializes for the blob. Please let me know if this needs to be addressed now, I could propose a fix. Or (preferably) let @thesofproject/google  address this in successive PR.

Comment (lkoenig) [reply src/audio/google/google_rtc_audio_processing.c:337]: The data should be whole and not altered during the call of `GoogleRtcAudioProcessingReconfigure` and `GoogleRtcAudioProcessingParseSofConfigMessage`
As of today we do not have concurrent access protection enabled in the `GoogleRtcAudioProcessing`.

Comment (lyakh) [reply src/audio/google/google_rtc_audio_processing.c:337]: @singalsu at least I would note in the comment, that a race is possible and should be fixed or at least further investigated

**Review by lgirdwood — APPROVED** [review]

**Comment by cujomalainey** [comment]
Apologies for the delay here, I will look on monday, been a crazy week

**Comment by lgirdwood** [comment]
@singalsu still draft ?

**Comment by singalsu** [comment]
> @singalsu still draft ?

I think it's best to wait for merge of #8237 that is a larger change.

File: src/audio/google/google_rtc_audio_processing_mock.c
Comment (ranj063) [inline src/audio/google/google_rtc_audio_processing_mock.c:154]: @singalsu this works for the 2ch case where the PDM 0/1 are configured on the board but if only PDM 2/3 are connected, this won't work correctly right?

File: src/audio/google/google_rtc_audio_processing.c
Comment (btian1) [inline src/audio/google/google_rtc_audio_processing.c:11]: if this patch have multiple purpose(from commit message), I would suggest split out 1 or 2 patches.

File: src/audio/google/google_rtc_audio_processing_mock.c
Comment (btian1) [inline src/audio/google/google_rtc_audio_processing_mock.c:158]: is it a bug? since previous code does not have channels loop.

Comment (singalsu) [reply src/audio/google/google_rtc_audio_processing_mock.c:158]: The original code mixed only to first channel, with this loop we can mix to other output channels. But according to Ranjani's comment we should decide what to do with non-matching channels counts.

File: src/audio/google/google_rtc_audio_processing.c
Comment (cujomalainey) [reply src/audio/google/google_rtc_audio_processing.c:11]: yes please

File: src/audio/google/google_rtc_audio_processing_mock.c
Comment (cujomalainey) [reply src/audio/google/google_rtc_audio_processing_mock.c:154]: in an ideal world, we would reduce our channels to only the ones we care about so we don't need these silly configs and only have real channels in the data.

File: src/arch/host/configs/library_defconfig
Comment (cujomalainey) [inline src/arch/host/configs/library_defconfig:20]: you can just enable the stub config

Comment (singalsu) [reply src/arch/host/configs/library_defconfig:20]: Thanks, helps with two other conversions also, done!

File: src/audio/google/google_rtc_audio_processing.c
Comment (singalsu) [reply src/audio/google/google_rtc_audio_processing.c:519]: Good catch, thanks!

File: src/audio/google/google_rtc_audio_processing_mock.c
Comment (singalsu) [reply src/audio/google/google_rtc_audio_processing_mock.c:154]: I changed it to version where every output channel has a mix of mic and reference. Hope it's good for testing now.

File: src/arch/host/configs/library_defconfig
Comment (btian1) [inline src/arch/host/configs/library_defconfig:6]: is google RTC processing is a common feature or platform specific feature? 
if it is platform specific, I  would suggest put it to platform config.

Comment (singalsu) [reply src/arch/host/configs/library_defconfig:6]: The name of this config file is perhaps misleading, though there's "host" in the path. This is for testbench and it's safe to have every component included with it. Testbench build is a good catcher in CI for build errors since we have many components with "default n" in kconfig those might be missed in normal platform builds.

File: src/audio/google/google_rtc_audio_processing.c
Comment (singalsu) [reply src/audio/google/google_rtc_audio_processing.c:517]: It's because module adapter provides the input and ouput streams as arrays of pointers to process(). I use the index to access the right stream. This also eases the transition to source/sink api in pipeline2.0 with better abstraction where process() receives array of sof_source and sof_sink pointers. It let's us optimize further the pipeline buffers. I'm not sure keeping the old way would prevent that but I felt this is cleaner approach.

What do you think @marcinszkudlinski , was this the right thing to do?

File: src/audio/google/google_rtc_audio_processing.c
Comment (singalsu) [reply src/audio/google/google_rtc_audio_processing.c:71]: Sorry, which pointer? I'm removing above the comp_buffet pointers and I think the rest of pointers in google_rtc_audio_processing_comp_data are still used. But I may well miss here something.

File: src/audio/google/google_rtc_audio_processing.c
Comment (singalsu) [reply src/audio/google/google_rtc_audio_processing.c:683]: Yes, not very clean API -- frames as input and bytes as output. But as far as I understand the design is from reference firmware. Did you mean a similar helper function but for this multiple sources/sinks case? This component will be converted to source/sink API after this PR for DP mode so we could review and possibly make improvements to api there.

FYI @marcinszkudlinski

**Comment by singalsu** [comment]
Marked this as DNM - I'm not sure if this PR step is necessary in converting this to DP. But please continue reviewing and sharing your thoughts so the next step can be better.

File: src/audio/google/google_rtc_audio_processing.c
Comment (cujomalainey) [reply src/audio/google/google_rtc_audio_processing.c:71]: nvm i got my the buffers confused between the alloced ones versus the SOF ones, apologies

File: src/audio/google/google_rtc_audio_processing.c
Comment (cujomalainey) [reply src/audio/google/google_rtc_audio_processing.c:683]: Yes, i think breaking into this struct is probably not the safest thing for API purposes on the buffer, we should fix the api or make a new call.

**Comment by lgirdwood** [comment]
> Marked this as DNM - I'm not sure if this PR step is necessary in converting this to DP. But please continue reviewing and sharing your thoughts so the next step can be better.

All modules should use the module API as component API is deprecated and will gradually be removed/refactored.i.e. we only have 1 API for modules.

**Comment by marcinszkudlinski** [comment]
Please proceed with merging
I'm finishing with changes to sink/src interface and DP processing of the module, all my changes have been based on top of this PR
So pls proceed and merge it once ready

**Comment by singalsu** [comment]
@cujomalainey Is this version good for merge? It will be a short time version only. Work by @marcinszkudlinski for DP schduling is now done and needs this patch as base.

**Review by lkoenig — APPROVED** [review]
The change looks good: I haven't look into too much details.
Did you run it with the mock  an d got good audio?

File: src/arch/host/configs/library_defconfig
Comment (lkoenig) [reply src/arch/host/configs/library_defconfig:6]: Don't you want to have the mock config as well here ?

**Review by lkoenig — APPROVED** [review]
Nice change so we can some at least build coverage.

File: src/audio/google/google_rtc_audio_processing_mock.c
Comment (lkoenig) [inline src/audio/google/google_rtc_audio_processing_mock.c:165]: You might mix several time same channel like is output is 2 and input is one channel you'll get 2 times the first channel.
```
 out[io] = sat_int16(
            (im < state->num_capture_channels ? (int32_t)mic[im++] : 0) + 
            (ir < state->num_reference_channles ? ref[ir++] : 0)
            );
```
Might be more appropriate (sorry for the vey poor codestyle).

**Review by lkoenig — CHANGES_REQUESTED** [review]

File: src/audio/google/google_rtc_audio_processing_mock.c
Comment (singalsu) [reply src/audio/google/google_rtc_audio_processing_mock.c:165]: We don't really know what's best for testing with mock-up. Sure, I can update it to be like this.

Comment (singalsu) [reply src/audio/google/google_rtc_audio_processing_mock.c]: @lkoenig Thanks, I included your suggestion as such, it works well!

**Review by lkoenig — APPROVED** [review]

**Comment by cujomalainey** [comment]
Sorry been busy internally, I trust @lkoenig 's review

**Review by lyakh — APPROVED** [review]

**Review by lgirdwood — APPROVED** [review]

**Review by cujomalainey — APPROVED** [review]

**Pull request merged**
